projects/driver_controls/test/test_can_output.c

```diff
@@ -0,0 +1,315 @@
+  TEST_ASSERT_TRUE(event_arbiter_process_event(&e));
+
+  TEST_ASSERT_EQUAL(CAN_OUTPUT_DEVICE_TURN_SIGNAL, s_can_output.id);
+  TEST_ASSERT_EQUAL(TURN_SIGNAL_FSM_STATE_LEFT_SIGNAL, s_can_output.state);
+  TEST_ASSERT_EQUAL(0, s_can_output.data);
+
+  // Test that a right turn signal generates the correct event
+  e.id = INPUT_EVENT_TURN_SIGNAL_RIGHT;
+  TEST_ASSERT_TRUE(event_arbiter_process_event(&e));
+
+  TEST_ASSERT_EQUAL(CAN_OUTPUT_DEVICE_TURN_SIGNAL, s_can_output.id);
+  TEST_ASSERT_EQUAL(TURN_SIGNAL_FSM_STATE_RIGHT_SIGNAL, s_can_output.state);
+  TEST_ASSERT_EQUAL(0, s_can_output.data);
+
+  // Test that turning the signals off generates the correct event
+  e.id = INPUT_EVENT_TURN_SIGNAL_NONE;
+  TEST_ASSERT_TRUE(event_arbiter_process_event(&e));
+
+  TEST_ASSERT_EQUAL(CAN_OUTPUT_DEVICE_TURN_SIGNAL, s_can_output.id);
+  TEST_ASSERT_EQUAL(TURN_SIGNAL_FSM_STATE_NO_SIGNAL, s_can_output.state);
+  TEST_ASSERT_EQUAL(0, s_can_output.data);
+}
+
+void test_can_fsm_direction() {
+  Event e = { 0 };
+
+  // Setup for the direction selector to be used
+  prv_toggle_power(true);
+
+  prv_toggle_mech_brake(true);
+
+  // Test that a forward shift generates the correct event
+  e.id = INPUT_EVENT_DIRECTION_SELECTOR_DRIVE;
+  TEST_ASSERT_TRUE(event_arbiter_process_event(&e));
+
+  TEST_ASSERT_EQUAL(CAN_OUTPUT_DEVICE_DIRECTION_SELECTOR, s_can_output.id);
+  TEST_ASSERT_EQUAL(DIRECTION_FSM_STATE_FORWARD, s_can_output.state);
+  TEST_ASSERT_EQUAL(0, s_can_output.data);
+
+  // Test that a reverse shift generates the correct event
+  e.id = INPUT_EVENT_DIRECTION_SELECTOR_REVERSE;
+  TEST_ASSERT_TRUE(event_arbiter_process_event(&e));
+
+  TEST_ASSERT_EQUAL(CAN_OUTPUT_DEVICE_DIRECTION_SELECTOR, s_can_output.id);
+  TEST_ASSERT_EQUAL(DIRECTION_FSM_STATE_REVERSE, s_can_output.state);
+  TEST_ASSERT_EQUAL(0, s_can_output.data);
+
+  // Test that a neutral shift generates the correct event
+  e.id = INPUT_EVENT_DIRECTION_SELECTOR_NEUTRAL;
+  TEST_ASSERT_TRUE(event_arbiter_process_event(&e));
+
+  TEST_ASSERT_EQUAL(CAN_OUTPUT_DEVICE_DIRECTION_SELECTOR, s_can_output.id);
+  TEST_ASSERT_EQUAL(DIRECTION_FSM_STATE_NEUTRAL, s_can_output.state);
+  TEST_ASSERT_EQUAL(0, s_can_output.data);
+}
+
+void test_can_fsm_pedal() {
+  Event e = {.data = 0 };
+
+  // Setup for the pedals to be used
+  prv_toggle_power(true);
+
+  prv_toggle_mech_brake(true);
+
+  e.id = INPUT_EVENT_DIRECTION_SELECTOR_DRIVE;
+  TEST_ASSERT_TRUE(event_arbiter_process_event(&e));
+
+  prv_toggle_mech_brake(false);
+
+  // Test that coasting generates the correct event
+  e = (Event){.id = INPUT_EVENT_PEDAL_COAST, .data = 0xdef };
+  TEST_ASSERT_TRUE(event_arbiter_process_event(&e));
+
+  TEST_ASSERT_EQUAL(CAN_OUTPUT_DEVICE_PEDAL, s_can_output.id);
+  TEST_ASSERT_EQUAL(PEDAL_FSM_STATE_COAST, s_can_output.state);
+  TEST_ASSERT_EQUAL(0xdef, s_can_output.data);
+
+  // Test that pressing the gas generates the correct event
+  e = (Event){.id = INPUT_EVENT_PEDAL_PRESSED, .data = 0xdef };
```

Comment:
Since the event data no longer gets clipped, it might be nice to verify that a full 16-bit integer would work.

Reply:
Done